server/src/main/scala/services/graphstages/ArrivalSplitsGraphStage.scala

```diff
@@ -0,0 +1,307 @@
+      }
+
+      super.preStart()
+    }
+
+    setHandler(outArrivalsWithSplits, new OutHandler {
+      override def onPull(): Unit = {
+        log.debug(s"arrivalsWithSplitsOut onPull called")
+        pushStateIfReady()
+        pullAll()
+      }
+    })
+
+    setHandler(inArrivalsDiff, new InHandler {
+      override def onPush(): Unit = {
+        log.debug(s"inFlights onPush called")
+        val arrivalsDiff = grab(inArrivalsDiff)
+
+        log.info(s"Grabbed ${arrivalsDiff.toUpdate.size} updates, ${arrivalsDiff.toRemove.size} removals")
+
+        val updatedFlights = purgeExpiredArrivals(updateFlightsFromIncoming(arrivalsDiff, flightsByFlightId))
+        log.info(s"We now have ${updatedFlights.size} arrivals")
+        arrivalsWithSplitsDiff = updatedFlights.values.toSet -- flightsByFlightId.values.toSet
+        log.info(s"arrivalsWithSplitsDiff: $arrivalsWithSplitsDiff")
+        flightsByFlightId = updatedFlights
+
+        pushStateIfReady()
+        pullAll()
+      }
+    })
+
+    setHandler(inManifests, new InHandler {
+      override def onPush(): Unit = {
+        log.debug(s"inSplits onPush called")
+        val vms = grab(inManifests)
+
+        log.info(s"Grabbed ${vms.manifests.size} manifests")
+        val updatedFlights = updateFlightsWithManifests(vms.manifests, flightsByFlightId)
+        log.info(s"We now have ${updatedFlights.size} arrivals")
+        arrivalsWithSplitsDiff = updatedFlights.values.toSet -- flightsByFlightId.values.toSet
+        log.info(s"arrivalsWithSplitsDiff: $arrivalsWithSplitsDiff")
+        flightsByFlightId = updatedFlights
+        manifestsBuffer = purgeExpiredManifests(manifestsBuffer)
+
+        pushStateIfReady()
+        pullAll()
+      }
+    })
+
+    setHandler(inSplitsPredictions, new InHandler {
+      override def onPush(): Unit = {
+        log.debug(s"inSplitsPredictions onPush called")
+        val predictions = grab(inSplitsPredictions)
+
+        log.info(s"Grabbed ${predictions.length} predictions")
+        val updatedFlights = addPredictions(predictions, flightsByFlightId)
+        arrivalsWithSplitsDiff = updatedFlights.values.toSet -- flightsByFlightId.values.toSet
+        flightsByFlightId = updatedFlights
+
+        pushStateIfReady()
+        pullAll()
+      }
+    })
+
+    def pullAll(): Unit = {
+      List(inManifests, inArrivalsDiff, inSplitsPredictions).foreach(in => if (!hasBeenPulled(in)) {
+        log.info(s"Pulling ${in.toString}")
+        pull(in)
+      })
+    }
+
+    def addPredictions(predictions: Seq[(Arrival, Option[ApiSplits])], flightsById: Map[Int, ApiFlightWithSplits]): Map[Int, ApiFlightWithSplits] = {
+      predictions
+        .collect {
+          case (arrival, Some(splits)) => (arrival, splits)
+        }
+        .foldLeft(flightsById) {
+          case (existingFlightsByFlightId, (arrivalForPrediction, predictedSplits)) =>
+            existingFlightsByFlightId.find {
```

Comment:
this is kind of tricky to follow, perhaps it could be broken into some functions?

Reply:
This whole function, not just this line, sorry.

Reply:
yep i agree